azure-proto-core/Resources/Plan.cs

```diff
@@ -27,82 +27,17 @@ public int CompareTo(Plan other)
                 return 1;
             }
 
-            if (Name == null && other.Name != null)
+            int compareResult = 0;
+            if ((compareResult = string.Compare(Name, other.Name)) == 0 &&
+                (compareResult = string.Compare(Product, other.Product)) == 0 &&
+                (compareResult = string.Compare(PromotionCode, other.PromotionCode)) == 0 &&
+                (compareResult = string.Compare(Publisher, other.Publisher)) == 0 &&
+                (compareResult = string.Compare(Version, other.Version)) == 0)
             {
-                return -1;
+                return 0;
             }
 
-            if (Name != null)
-            {
-                var compareNameResult = Name.CompareTo(other.Name);
-
-                if (compareNameResult != 0)
-                {
-                    return compareNameResult;
-                }
-            }
-
-            if (Product == null && other.Product != null)
-            {
-                return -1;
-            }
-
-            if (Product != null)
-            {
-                var compareProductResult = Product.CompareTo(other.Product);
-
-                if (compareProductResult != 0)
-                {
-                    return compareProductResult;
-                }
-            }
-
-            if (PromotionCode == null && other.PromotionCode != null)
-            {
-                return -1;
-            }
-
-            if (PromotionCode != null)
-            {
-                var comparePromotionCodeResult = PromotionCode.CompareTo(other.PromotionCode);
-
-                if (comparePromotionCodeResult != 0)
-                {
-                    return comparePromotionCodeResult;
-                }
-            }
-
-            if (Publisher == null && other.Publisher != null)
-            {
-                return -1;
-            }
-
-            if (Publisher != null)
-            {
-                var comparePublisherResult = Publisher.CompareTo(other.Publisher);
-
-                if (comparePublisherResult != 0)
-                {
-                    return comparePublisherResult;
-                }
-            }
-
-            if (Version == null && other.Version != null)
-            {
-                return -1;
-            }
-
-            if (Version != null)
-            {
-                var compareVersionResult = Version.CompareTo(other.Version);
-
-                if (compareVersionResult != 0)
-                {
-                    return compareVersionResult;
-                }
-            }
-
-            return 0;
+            return compareResult;
```

Comment:
I believe this removes the ordering that we had in place.  Before Name had priority for comparison over Version as an example, so if i had 2 objects A and B and A.Name > B.Name but A.Version < B.Version compareTo will now return -1 instead of 1 since the last value in compareResult will be A.Version.CompareTo(B.Version).  This gets worse in the following scenario
A.Name > B.Name && A.Version == B.Version In this case again the expected value is 1, but the way the code is written the actual value will be 0.

Please add the example tests cases I have given here to prove the assertion either way.

Reply:
A and B A.Name < B.Name && A.Version > B.Version expected -1

Reply:
I tested the three scenarios, they all can get the expected value, The updates in this PR doesn't remove the ordering, it will compare Name first, if not equals will jump out the if statement and return `compareResult`, same as other properties.